### PR TITLE
PRP-227 - Magical routing!

### PR DESCRIPTION
**Pull request by microwavenby**
## Ticket

https://wicmtdp.atlassian.net/browse/PRP-227

## Changes
* URLs are validated to check that there is a valid Local Agency in them
* Excess / incorrect paths are redirected to real paths
* Invalid sessions redirect to the same clinic (if valid)
* Navigating to / bumps the individual to `/gallatin/recertify` (the first agency in the database)
* Added tests for redirect logic
* Added playwright tests for redirect logic, cookie validation
* Updated seed script to support atomic runs (rerunning it won't hurt)
* Removed hardcoded clinic string from cookie parser
* Centralized all logic for routing in utils/redirect.ts

## Testing
1.  run `npm run dev`
2. navigate to `http://localhost:3000/bogus/url/somewhere`
3. See `http://localhost:3000/gallatin/recertify` in your browser

More exciting / advanced tests are available. Real routes should still work for valid clinics, for example
(http://localhost:3000/gallatin/recertify/intro) or (http://localhost:3000/missoula/recertify/intro)

### CODE REVIEW

File: participant/prisma/seed.ts [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
+  const existingAgencyRecords = await prisma.localAgency.count();
+  if (existingAgencyRecords != agencyData.length) {
+    await prisma.localAgency.createMany({
+      data: agencyData,
+      skipDuplicates: true,
+    });
+    console.log(`Database has been seeded. 🌱`);
+  } else {
+    console.log(
+      `Skipping localAgency seed; found ${existingAgencyRecords} agencies 🪴`
```
Comment (rocketnova) [inline participant/prisma/seed.ts:29]: 🥺🪴

File: participant/app/cookies.server.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import { findSubmission, upsertSubmission } from "./utils/db.server";
+import { validRoute } from "./utils/redirect";
+const MAX_SESSION_SECONDS = Number(process.env.MAX_SESSION_SECONDS) || 1800;
+
+type ParticipantCookieContents = {
+  submissionID?: string;
+};
+
+// This should be secure: true, and have secrets in prod (probably)
+export const ParticipantCookie = createCookie("prp-recertification-form");
```
Comment (rocketnova) [inline participant/app/cookies.server.ts:14]: Did this change from #31 because setting it to `secure: true` breaks local development?

File: participant/app/cookies.server.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  }
+  console.log(`Generating ${submissionID}`);
+  cookie.submissionID = submissionID;
+  console.log(
+    `Creating Submission record in database for ${submissionID} and agency ${urlId}`
+  );
+  await upsertSubmission(submissionID, urlId);
+  if (forceRedirect) {
+    const redirectTarget = await validRoute(request, params, true);
+    if (redirectTarget) {
```
Comment (rocketnova) [inline participant/app/cookies.server.ts:74]: It took me a minute to figure out that the case where `forceRedirect == true` and `redirectTarget == null` is not 💣 but actually a no-op. Could you include a comment here to indicate that this is ok and when this might happen?

File: participant/app/utils/db.server.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+export const findLocalAgency = async (urlId: string) => {
+  return await db.localAgency.findUnique({
+    where: {
+      urlId: urlId,
+    },
+  });
+};
+
+export const firstLocalAgency = async () => {
```
Comment (rocketnova) [inline participant/app/utils/db.server.ts:48]: Possibly excessive, but might be good to have a unit test for this to catch any possible future regressions.

File: participant/app/root.tsx [hunk 5]
```diff
@@ -48,10 +50,17 @@ export function links() {
 
 type LoaderData = { locale: string; demoMode: string; missingData: string };
 
-export const loader: LoaderFunction = async ({ request }) => {
+export const loader: LoaderFunction = async ({ request, params }) => {
+  const redirectTarget = await validRoute(request, params);
+  if (redirectTarget) {
+    console.log(`Redirecting to first agency ${redirectTarget}`);
```
Comment (rocketnova) [inline participant/app/root.tsx]: This isn't always the first agency, right? For example, this condition would be true in the case where the user goes to `/${mockAgency.urlId}/recertify/somewhere/else`, right?

File: participant/app/utils/redirect.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import type { Params } from "@remix-run/react";
+import { findLocalAgency, firstLocalAgency } from "./db.server";
+
+export const validRoute = async (
+  request: Request,
+  params?: Params<string>,
+  findRoot: boolean = false
```
Comment (rocketnova) [inline participant/app/utils/redirect.ts]: Nit: Could we call this something other than "root"? I find it a little bit confusing because it's not strictly `/`, but rather `$localAgency/recertify`. Maybe `findBase`? I like that you referred to it as the base url in the comments for the e2e test.

File: participant/tests/utils/redirect.test.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  const request = makeRequest(`/${mockAgency.urlId}/recertify/somewhere/else`);
+  const target = await validRoute(
+    request,
+    { localAgency: mockAgency.urlId },
+    true
+  );
+  expect(target).toBe(`/${mockAgency.urlId}/recertify`);
+});
+
+it("will ignore the url if params are present", async () => {
```
Comment (rocketnova) [inline participant/tests/utils/redirect.test.ts:63]: Would this condition only happen if there were a bug in Remix and somehow the `params` diverged from the `request` and in that weird edge case, we would ignore the weird `request` and use the `params` instead?

File: participant/app/utils/redirect.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    if (firstSegment) {
+      const agency = await findLocalAgency(firstSegment);
+      if (agency) {
+        redirectToFirstAgency = false;
+        validAgencyId = agency.urlId;
+      }
+    }
+    if (!validAgencyId) {
+      redirectToFirstAgency = true;
+    }
```
Comment (rocketnova) [inline participant/app/utils/redirect.ts:31]: I think you don't have a unit test for this edge case where the path is something like `/bogusAgency`?

File: participant/app/routes/$.tsx [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { redirect } from "@remix-run/server-runtime";
+import type { LoaderFunction } from "@remix-run/server-runtime";
+import { validRoute } from "~/utils/redirect";
+
+export const loader: LoaderFunction = async ({ request, params }) => {
+  const url = new URL(request.url);
+  if (url.pathname.includes("favicon.ico")) {
+    return null;
```
Comment (rocketnova) [inline participant/app/routes/$.tsx]: Why do we need this? Shouldn't Remix be able to handle file requests?

**Review by rocketnova — COMMENTED** [review]
This is looking so good! A few inline questions and comments.

File: participant/app/cookies.server.ts (hunk 3 above)
Comment (microwavenby) [reply participant/app/cookies.server.ts:74]: It should only happen if the database has no agency in it to direct to first agency.. maybe?

I think probably we should have the `validRoute` throw you somewhere on the `/error/routingError` vibe instead

File: participant/tests/utils/redirect.test.ts (hunk 7 above)
Comment (microwavenby) [reply participant/tests/utils/redirect.test.ts:63]: Yea, that's right. It's definitely an unexpected edge case

File: participant/app/routes/$.tsx (hunk 9 above)
Comment (microwavenby) [reply participant/app/routes/$.tsx]: I don't know why we're getting requests for `/gallatin/recertify/favicon.ico` -- but because we are, those requests are hitting the splat route. 
Requests for `/favicon.ico` work perfectly fine.

File: participant/app/utils/db.server.ts (hunk 4 above)
Comment (microwavenby) [reply participant/app/utils/db.server.ts:48]: I'll write one, but as it relies on mocking, it's only going to really test that `firstLocalAgency` calls `prisma.localAgency.findFirst`

**Review by rocketnova — APPROVED** [review]
This looks really great. Thank you!
